Add tests for System.Security.AccessControl ACE classes

ianhays opened the pull request:
There doesn't seem to be any tests for System.Security.AccessControl in open. I decided to port these tests in chunks instead of all at once to ease reviewing and get feedback on structure before it was difficult to modify. 

The tests in ToF are <i>very</i> incompatible with Xunit so after wasting a bunch of time trying to get them running in Open, I opted to rewrite them instead. The structure of the tests is different than it was, but the MemberData we're testing with is the same.

Progress towards #6922.

@stephentoub

stephentoub commented:
cc: @bartonjs

ianhays commented:
```
Test Result (2 failures / +2)
System.Net.Security.Tests.ClientAsyncAuthenticateTest.ClientAsyncAuthenticate_AllServerVsIndividualClientSupportedProtocols_Success(clientProtocol: Tls)
System.Net.Security.Tests.SslStreamStreamToStreamTest.SslStream_StreamToStream_LargeWrites_Sync_Success(randomizedData: False)

Test Result (1 failure / +1)
System.Linq.Parallel.Tests.GetEnumeratorTests.GetEnumerator_DisposeAsynchronously(synchronousMerge: False)
```

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+            Assert.Throws<ArgumentNullException>("binaryForm", () => CustomAce.CreateFromBinaryForm(null, 1));
+            Assert.Throws<ArgumentOutOfRangeException>("offset", () => CustomAce.CreateFromBinaryForm(new byte[1], -1));
+            Assert.Throws<ArgumentException>("binaryForm", () => CustomAce.CreateFromBinaryForm(new byte[ace.BinaryLength + 1], 2));
+            Assert.Throws<ArgumentException>("binaryForm", () => CustomAce.CreateFromBinaryForm(new byte[ace.BinaryLength], 1));
```
CustomAce.CreateFromBinaryForm and GenericAce.CreateFromBinaryForm and all other CreateFromBinaryForms are the same.

Since there isn't actually a CustomAce.CreateFromBinaryForm this should be changed to call it where it's really defined, GenericAce.CreateFromBinaryForm.  And then a lot of tests might disappear.

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+        {
+            yield return CustomAce_CreateTestData(19, 0, 4, 0);
+            yield return CustomAce_CreateTestData(18, 1, 8, 0);
+            yield return CustomAce_CreateTestData(17, 1, 4, 0);
```
Do any of these numbers have meaning?

ianhays replied:
The old tests stored these values in text files then read them into their tests with a custom reader. I used those same values but modfied the format to work as memberdata instead.

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+            CustomAce ace = new CustomAce(type, flags, opaque);
+            Assert.Equal(type, ace.AceType);
+            Assert.Equal(flags, ace.AceFlags);
+            Assert.Equal(opaque, ace.GetOpaque());
```
This would be more valueable if opaque weren't always an array of zeroes.

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => new CustomAce((AceType)16, (AceFlags)15, new byte[1]));
+            Assert.Throws<ArgumentOutOfRangeException>(() => new CustomAce((AceType)19, (AceFlags)1, new byte[1]));
+            Assert.Throws<ArgumentOutOfRangeException>(() => new CustomAce((AceType)0, (AceFlags)2, new byte[4]));
```
Why are any of these invalid?  Is there value in further checking the parameter name of the AOORE?

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+
+            foreach (AceType type in Enum.GetValues(typeof(AceType)))
+            {
+                Assert.Throws<ArgumentOutOfRangeException>(() => new CustomAce(type, (AceFlags)1, new byte[4]));
```
Why 1? And 4?

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+            yield return CompoundAce_CreateTestData(15, 1, 4, "S-1-5-11", 0);
+            yield return CompoundAce_CreateTestData(0, 1, 1, "S-1-5-11", 0);
+            yield return CompoundAce_CreateTestData(0, 1, 1, "S-1-5-11", 0);
+            yield return CompoundAce_CreateTestData(0, 1, 1, "S-1-5-11", 0);
```
Again, lots of magic numbers, and I have no idea why any of them are what they are.  It's not really clear to me what we're testing.

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+        }
+    }
+
+    public class CompoundAce_Tests : KnownAce_Tests
```
I know that the product code has like 85 types in one file, but can we break these into multiple files?

ianhays replied:
Yup, sure thing.

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+            Assert.Equal(sid, ace.SecurityIdentifier);
+            Assert.Equal(opaque, ace.GetOpaque());
+            Assert.Equal(qualifier, ace.AceQualifier);
+            Assert.Equal(isCallback, ace.IsCallback);
```
These blocks (in all the Ace_Test types) feel like they just want to be a helper assert method, that way you can hierarchy cascade them.

ianhays replied:
I think I already wrote a ValidateGenericAce method for that purpose but decided against using it for some reason. I'll put it back in.

bartonjs commented on src/System.Security.AccessControl/tests/Ace.Tests.cs:
```diff
@@ -0,0 +1,482 @@
+            GenericAce_VerifyAces(expectedAce, resultAce);
+        }
+    }
+}
```
Tests I didn't see:

* GenericAce.Equals
    * It looks like if you made a `TestAce : GenericAce` and it had the same binary form as another Ace type that Equals would match even though the types don't.  If that's true in core (and desktop) we should have a test to make sure we don't ever break that.
* GetHashCode
    * Shouldn't really be tested for a known value, but that two distinct instances which .Equals the same GetHashCode the same.  It's hard to codify that "different should be different" because of how hashing works, but feel free to make some sort of generalization :)
* op_Equals
    * Nulls, `a == a`, value equals
* op_NotEquals
    * Pretty easily bolted on to the previous test

After GenericAce I stopped scrolling through looking to see if there were any other public members not being utilized.

ianhays replied:
If a test isn't in the PR it's because we didn't have a ToF equivalent. I won't be adding anything new in this PR, only porting what we already have, rewriting it to play nice with Xunit, and making sure there aren't any breaking failures.

That's also why there are a bunch of hard-coded values throughout the tests - because that's what was in the original tests. Like you, I don't know what most of the values mean; I'm just parroting them from ToF for the port. There are also a bunch of gaps in those tests, but that's out-of-scope for this push.

That in mind, I really should have made dotnet-bot author the initial commit and then added the playnice features on top of that. I'll rebase to clear that up.

ianhays commented:
Rebased on top of dotnet-bot and split up the ACE tests to a few different classes.

I've got another PR ready to put on top of this one that ports the rest of the tests for AccessControl. Those I decided to not rewrite, though I did write some scripts to remove extra crap and convert validation checks to xunit style.

ianhays commented:
@bartonjs Are you fine if I merge this and open another issue for fixing them e.g. looking back through them, cleaning them up, and adding new tests?

I'd prefer to keep the port separate from any actual improvements.

bartonjs commented:
@ianhays You'll need to update the project.json again, or it'll be a build break (sorry for the delays).  Then it'd be great if you squashed your commits (as opposed to dotnet-bot's) down into one; so there's just "copy code" and "xunit-ify the code".  Then anything that you're not feeling inspired to clean up while you're in there just create an issue for it.

So, post build number reference update and partial squash, LGTM.

ianhays commented:
Thanks for the heads-up on the project.json! I updated it and squashed down to two commits. I'll merge when CI is green.

ellismg commented:
Both windows legs failed with errors like this:

> 15:57:28 D:\j\workspace\windows_nt_de---06b7984d\Tools\tests.targets(145,5): warning MSB3073: The command "D:\j\workspace\windows_nt_de---06b7984d\bin/tests/Windows_NT.AnyCPU.Debug/System.Security.AccessControl.Tests/dnxcore50/CoreRun.exe xunit.console.netcore.exe System.Security.AccessControl.Tests.dll  -xml testResults.xml -notrait Benchmark=true  -notrait category=OuterLoop -notrait category=failing -notrait category=nonwindowstests " exited with code 9009. [D:\j\workspace\windows_nt_de---06b7984d\src\System.Security.AccessControl\tests\System.Security.AccessControl.Tests.csproj]

It looks like a real thing.

bartonjs commented:
Possibly this is also relevant:

> 16:00:27 C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1819,5): warning MSB3277: Found conflicts between different versions of the same dependent assembly that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [D:\j\workspace\windows_nt_de---06b7984d\src\System.Security.AccessControl\tests\System.Security.AccessControl.Tests.csproj]

ianhays commented:
Those were legit issues. The Release tests were failing because I didn't have:

 ```
     "test-runtime": {
      "target": "project",
      "exclude": "compile"
    }
```
 in my project.json.

ianhays commented:
test Innerloop OSX Debug Build and Test please
test Innerloop OSX Release Build and Test please

ianhays commented:
test  Innerloop Windows_NT Debug Build and Test please
